R/topTaxa.R

```diff
@@ -0,0 +1,90 @@
+#' Returns top N \dQuote{FeatureID}
+#'
+#' Basic extractor function for extracting top N \dQuote{FeatureID}
+#' from assay for a user specified \dQuote{SampleID} in a
+#' \code{\link[=SummarizedExperiment-class]{SummarizedExperiment}}
+#' object.
+#'
+#' @param x A
+#'  \code{\link[=SummarizedExperiment-class]{SummarizedExperiment}} object.
+#'
+#' @param top Numeric value, how many top taxa to return. Default return top five taxa.
+#'
+#' @param method Specify the method to determine top taxa. Either sum, mean or median.
+#'               Default is 'mean'
+#'
+#' @param abund_values a \code{character} value to select an
+#'   \code{\link[SummarizedExperiment:SummarizedExperiment-class]{assayNames}}
+#'
+#' @importFrom DelayedMatrixStats rowSums2 rowMeans2 rowMedians
+#'
+#' @return A vector of the top N \dQuote{FeatureID}
+#' @name topTaxa
+#'
+#' @author
+#' Sudarshan A. Shetty
+#'
+#' @examples
+#' data(GlobalPatterns, package = "MicrobiomeExperiment")
+#'
+#' top_taxa <- topTaxa(GlobalPatterns,
+#'                     method="mean",
+#'                     top=5,
+#'                     abund_values="counts")
+#'
+NULL
+
+#' @rdname topTaxa
+#'
+#' @export
+
+setGeneric("topTaxa", signature = "x",
+           function(x, top, method, abund_values)
+               standardGeneric("topTaxa"))
+
+.check_max_taxa <- function(x, top, abund_values){
+
+    if(top > nrow(assay(x,abund_values))){
+        stop("top value is greater than maximum number of taxa in assay", call. = FALSE)
+    }
+
+}
+
+#' @rdname topTaxa
+#'
+#' @export
+#'
+setMethod("topTaxa", signature = c(x = "SummarizedExperiment"),
+          function(x, top=5, method="mean", abund_values = "counts"){
```

Comment:
the value of mean is not sanitised. Setting `mean = "whatever"` will crash this whole function.

Best practice would be to add the generic like this:
```
setGeneric("topTaxa", signature = "x",
            function(x, ...)
                standardGeneric("topTaxa"))
```

and implement the code like this:

```
setMethod("topTaxa", signature = c(x = "SummarizedExperiment"),
           function(x, top = 5L, method=c("mean","sum","median"), abund_values = "counts"){
                method <- match.arg(method)
...
})
```

That way, the generic for `topTaxa` can be reused, if other definition of `topTaxa` need to be implemented. (See comment above wrapping `getTopHVGs`)

However, if you would like to stick to the current implementation more closely you could do this

```
setGeneric("topTaxa", signature = "x",
            function(x, top=5L, method=c("mean","sum","median"), abund_values = "counts")
                standardGeneric("topTaxa"))
```

and implement the code like this:

```
setMethod("topTaxa", signature = c(x = "SummarizedExperiment"),
           function(x, top = 5L, method=c("mean","sum","median"), abund_values = "counts"){
                method <- match.arg(method, c("mean","sum","median"))
...
})
```
Please note that default values have to defined in the generic, if the argument is defined there. In the first example this is not the case and therefore the additional arguments are defined by `setMethod` with their defaults

For more details on `match.arg`, see `?match.arg`